File: doc/tutorial/nearestneighbor/code.cc
#include "code.h"
#include <math.h>

/**
 * return the distance between point0 and point1
 */
double distance_between_points(
    double x0, double y0, double z0, 
    double x1, double y1, double z1)
{
    double dx = x1 - x0;
    double dy = y1 - y0;
    double dz = z1 - z0;
    return sqrt(dx * dx  + dy * dy + dz * dz);
}

class DistanceAndIndex{
    
public:
    double r;
    int index;
    
    DistanceAndIndex():r(-1),index(-1) {}
    DistanceAndIndex(DistanceAndIndex & original):r(original.r),index(original.index){}
};

/**
 * Find the nearest neigbors of all the points (specified with
 * x, y and z).
 * 
 * Fills the n1, n2 and n3 arrays with the closest, second clostest
 * and third closests points.
 * 
 **/
int find_nearest_neighbors(int npoints, 
    double * x, double * y, double * z,
    int * n0, int * n1, int * n2)
{
    for(int i = 0; i < npoints; i++)
    {
        DistanceAndIndex found[3];
        
        for(int j = 0; j < npoints; j++)
        {
            if(j == i) {
                continue;
            }
            
            double x0 = x[i];
            double y0 = x[i];
            double z0 = x[i];
            
            double x1 = x[j];
            double y1 = x[j];
            double z1 = x[j];
            
            double r = distance_between_points(x0, y0, z0, x1, y1 , z1);
            
            for(int k = 0; k < 3; k++) {
                if(found[k].index == -1) {
                    found[k].index = j;
                    found[k].r = r;
                    break;
                } else {
                    if(r < found[k].r) {
                        for(int l = 2;l > k; l--) {
                            found[l] = found[l-1];
                        }
                        found[k].index = j;
                        found[k].r = r;
                        break;
                    }
                }
            }
            
        }
        
        n0[i] = found[0].index;
        n1[i] = found[1].index;
        n2[i] = found[2].index;
    }
    
    
    return 0;
}

File: doc/tutorial/nearestneighbor/code.f90
FUNCTION distance_between_points(x0,y0,z0,x1,y1,z1)
    DOUBLE PRECISION :: distance_between_points
    DOUBLE PRECISION :: x0,y0,z0,x1,y1,z1
    DOUBLE PRECISION :: dx,dy,dz
    dx = x1 - x0;
    dy = y1 - y0;
    dz = z1 - z0;
    distance_between_points = sqrt(dx * dx  + dy * dy + dz * dz);
END FUNCTION


FUNCTION find_nearest_neighbors(npoints,x,y,z,n0,n1,n2)
    INTEGER :: npoints
    INTEGER :: i,j, k, kk
    INTEGER :: find_nearest_neighbors
    DOUBLE PRECISION :: x(npoints),y(npoints),z(npoints)
    INTEGER :: n0(npoints),n1(npoints),n2(npoints)
    DOUBLE PRECISION :: r
    INTEGER :: nn_index(3)
    DOUBLE PRECISION :: nn_distance(3)
    
    DO i = 1, npoints, 1
        DO k = 1, 3
            nn_index(k) = 0
            nn_distance(k) = 0.0
        END DO
        
        DO j = 1, npoints, 1
            IF (i.NE.j) THEN
                r = distance_between_points(x(i), y(i), z(i), &
                    x(j), y(j), z(j))
                
                
                DO k = 1, 3
                    IF (nn_index(k).EQ.0) THEN
                        nn_index(k) = j
                        nn_distance(k) = r
                    ELSE
                        IF (r.LT.nn_distance(k)) THEN
                            DO kk = 3, k, -1
                                nn_index(kk) = nn_index(kk-1)
                                nn_distance(kk) = nn_distance(kk-1)
                            END DO
                            nn_index(k) = k
                            nn_distance(k) = r
                            EXIT
                        END IF
                    END IF
                    nn_distance(k) = 0.0
                END DO
            END IF
        END DO
        
        n0(i) = nn_index(1)
        n1(i) = nn_index(2)
        n2(i) = nn_index(3)
    END DO
    
    find_nearest_neighbors = 0
END FUNCTION
